test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java

```diff
@@ -327,12 +324,27 @@ public void test() throws IOException {
 
         try {
             for (ExecutableSection executableSection : testCandidate.getTestSection().getExecutableSections()) {
-                executeSection(executableSection);
+                try {
+                    executeSection(executableSection);
+                } catch (AssertionError | Exception e) {
+                    // Dump the stash on failure. Instead of dumping it in true json we escape `\n`s so stack traces are easier to read
+                    logger.info("Stash dump on test failure [{}]",
+                            Strings.toString(restTestExecutionContext.stash(), true, true)
+                                    .replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t"));
+                    throw e;
+                }
             }
         } finally {
             logger.debug("start teardown test [{}]", testCandidate.getTestPath());
             for (DoSection doSection : testCandidate.getTeardownSection().getDoSections()) {
-                executeSection(doSection);
+                try {
+                    executeSection(doSection);
+                } catch (AssertionError | Exception e) {
+                    logger.info("Stash dump on tear down failure [{}]",
+                            Strings.toString(restTestExecutionContext.stash(), true, true)
+                                    .replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t"));
+                    // Don't rethrow exception here, that could hide the exception caused by a test.
+                }
```

Comment:
should we rather handle this in the executeSection method so we don't have to repeat the same code three times?